File: 2018-go/four/nap.go
package four

type Nap struct {
	startMin int
	endMin   int
	diff     int
}

func NewNap(startMin int, endMin int) Nap {
	return Nap{startMin: startMin, endMin: endMin, diff: endMin - startMin}
}

func (n Nap) isAsleepAtMin(min int) bool {
	return min >= n.startMin && min < n.endMin
}

File: 2018-rust/src/four/nap.rs
use core::fmt::Debug;

pub struct Nap {
    start_min: usize,
    end_min: usize,
    pub diff: usize,
}

impl Debug for Nap {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{} mins from {} to {}", self.diff, self.start_min, self.end_min)
    }
}

impl Nap {
    pub fn new(start_min: usize, end_min: usize) -> Nap {
        Nap { start_min, end_min, diff: end_min - start_min }
    }
    
    pub fn is_asleep_at(&self, minute: usize) -> bool {
        return minute >= self.start_min && minute < self.end_min;
    }
}